### PR TITLE
Vcenter properties collector

### PR DESCRIPTION
**Pull request by durcar86**
a collector which get vcenter version info

`vrops_vcenter_properties{datacenter="NA-CA-1A",propkey="summary_version:6.5.0-11347054",vcenter="vc-a-0.cc.na-ca-1.cloud.sap"} 0.0
vrops_vcenter_properties{datacenter="NA-CA-1A",propkey="vc_fullname:VMware vCenter Server 6.5.0 build-11347054",vcenter="vc-a-0.cc.na-ca-1.cloud.sap"} 0.0`

### CODE REVIEW

**Comment by richardtief** [comment]
Before we review it, please adjust the test for your collectors. Go into the TestCollector and see how it works. You have to write the metrics you expect into the `metrics.yaml`

You can find information here: https://github.com/sapcc/vrops-exporter#test

**Comment by viennaa** [comment]
Additionally your exporter would need to be added in the Test:
https://github.com/sapcc/vrops-exporter/blob/master/tests/TestCollectors.py#L16-L19

**Comment by durcar86** [comment]
ok, let me look into this

**Comment by viennaa** [comment]
Still wants to be fixed :)
```Traceback (most recent call last):
  File "tests/TestCollectors.py", line 79, in test_collector_metrics
    for propkey_pair in propkey_yaml[collector]['enum_metrics']:
KeyError: 'enum_metrics'```

**Comment by durcar86** [comment]
local debug looks ok 

`waiting for initial iteration
done: initial query
prometheus query port 9578

Testing HostSystemPropertiesCollector
waiting for initial iteration
done: initial query
prometheus query port 9579

Testing VcenterPropertiesCollector
waiting for initial iteration
done: initial query
prometheus query port 9580

Testing DatastoreStatsCollector
waiting for initial iteration
done: initial query
prometheus query port 9581

Testing SampleCollector
waiting for initial iteration
done: initial query
prometheus query port 9582
..
----------------------------------------------------------------------
Ran 2 tests in 100.379s

OK`

**Comment by viennaa** [comment]
Please resolve the conflicting files.

**Comment by richardtief** [comment]
The method for a simple query returns only `test_property`. 

https://github.com/durcar86/vrops-exporter/blob/9fdc6a488423dfa6b5b62f53b93e7f05123e9493/tests/TestCollectors.py#L64

File: collectors/VCenterPropertiesCollector.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+
+class VcenterPropertiesCollector(BaseCollector):
+    def __init__(self):
+        self.iteration = 0
+        while not self.iteration:
+            time.sleep(5)
+            self.get_iteration()
+            print("waiting for initial iteration")
+        print("done: initial query")
```
Comment (richardtief) [inline collectors/VCenterPropertiesCollector.py]: Use the new `wait_for_inventory_data()` method here after your update with the master branch.

File: collectors/VCenterPropertiesCollector.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        self.property_yaml = YamlRead('collectors/property.yaml').run()
+
+    def collect(self):
+        if os.environ['DEBUG'] >= '1':
+            print('vCenterPropertiesCollector starts with collecting the properties')
+
+        g = GaugeMetricFamily('vrops_vcenter_properties', 'testtest',
+                              labels=['datacenter', 'vcenter', 'propkey'])
```
Comment (richardtief) [inline collectors/VCenterPropertiesCollector.py]: According to our latest findings, it is useful to insert a method that describes the collector. Please adopt this also in your collectors. To do this, you need to initialize the GaugeMetrik object in the init method. Make sure that you reference this object everywhere.

```
def __init__(self):
         self.wait_for_inventory_data()
         self.property_yaml = YamlRead('collectors/property.yaml').run()
         self.g = GaugeMetricFamily('vrops_vcenter_properties', 'testtest',
                                   labels=['datacenter', 'vcenter',  'propkey'])

    def describe(self):
        yield self.g
```

File: collectors/VCenterPropertiesCollector.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+        for vc in self.get_vcenters():
+            target = self.vcenters[vc]['target']
+            target_vc = self.vcenters[vc]['name']
+            token = self.vcenters[vc]['token']
+            vc_uuid = self.vcenters[vc]['uuid']
+            for cl in self.get_clusters():
+                if self.clusters[cl]['vcenter'] == target_vc:
+                    dc_name = self.clusters[cl]['parent_dc_name']
```
Comment (richardtief) [inline collectors/VCenterPropertiesCollector.py]: A suggestion for improvement: (If you want the datacenter information in your metric)

```
        for dc in self.get_datacenters():
            target = self.datacenters[dc]['target']
            target_vc = self.datacenters[dc]['parent_vcenter_name']
            token = self.datacenters[dc]['token']
            vc_uuid = self.datacenters[dc]['parent_vcenter_uuid']
            dc_name = self.datacenters[dc]['name']
```


But you don't have to set `dc_name` and `target_vc`. You can insert `self.datacenters[dc]['name']` and `self.datacenters[dc]['parent_vcenter_name']` directly in your labels.

**Comment by durcar86** [comment]
> The method for a single resource returns only `test_property`.
> 
> https://github.com/durcar86/vrops-exporter/blob/9fdc6a488423dfa6b5b62f53b93e7f05123e9493/tests/TestCollectors.py#L64

Thanks Richard, i tried this and now TestCollectors.py is failing

======================================================================
FAIL: test_collector_metrics (__main__.TestCollectors)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/TestCollectors.py", line 163, in test_collector_metrics
    self.assertTrue(set(metrics).issubset(metrics_yaml_list),
AssertionError: False is not true : VcenterPropertiesCollector: metric not covered by testcase, probably missing in yaml
vrops_vcenter_properties{datacenter="datacenter1",propkey="summary_version: test_property",vcenter="vcenter1"}
vrops_vcenter_properties{datacenter="datacenter1",propkey="vc_fullname: test_property",vcenter="vcenter1"}

----------------------------------------------------------------------
Ran 2 tests in 55.260s

**Comment by durcar86** [comment]
closing this PR and I will open a new one to resolve the conflicting files

**Pull request closed without merging**